cpp/src/stream_compaction/stream_compaction.cu

```diff
@@ -108,4 +115,118 @@ gdf_column drop_nulls(gdf_column const &input) {
+        return thrust::get<0>(x) && thrust::get<1>(x); }
+}; 
+
+template<typename DerivedPolicy,
+         typename InputIterator,
+         typename OutputIterator,
+         typename BinaryPredicate>
+__host__ __device__
+  OutputIterator non_duplicates_copy(thrust::execution_policy<DerivedPolicy> &exec,
+                             InputIterator first,
+                             InputIterator last,
+                             OutputIterator output,
+                             BinaryPredicate binary_pred)
+{
+  thrust::detail::head_flags<InputIterator, BinaryPredicate> stencil1(first, last, binary_pred);
+  thrust::detail::tail_flags<InputIterator, BinaryPredicate> stencil2(first, last, binary_pred);
+  auto combined_stencil = thrust::make_transform_iterator(
+          thrust::make_zip_iterator(thrust::make_tuple(stencil1.begin(), stencil2.begin())),
+          tuple_reduced_and());
+  
+  using namespace thrust::placeholders;
+  
+  return thrust::copy_if(exec, first, last, combined_stencil, output, _1);
+} 
+
+rmm::device_vector<gdf_index_type>
+get_unique_ordered_indices(const cudf::table& key_columns,
+                           const duplicate_keep_option keep)
+{
+  gdf_size_type ncols = key_columns.num_columns();
+  gdf_size_type nrows = key_columns.num_rows();
+
+  // sort only indices
+  rmm::device_vector<gdf_size_type> sorted_indices(nrows);
+  gdf_context context;
+  gdf_column sorted_indices_col;
+  CUDF_TRY(gdf_column_view(&sorted_indices_col, (void*)(sorted_indices.data().get()),
+        nullptr, nrows, GDF_INT32));
+  CUDF_TRY(gdf_order_by(key_columns.begin(),
+        nullptr,
+        key_columns.num_columns(),
+        &sorted_indices_col,
+        &context));
+
+  // extract unique indices 
+  rmm::device_vector<gdf_index_type> unique_indices(nrows);
+
+  cudaStream_t stream;
+  cudaStreamCreate(&stream);
+  auto exec = rmm::exec_policy(stream)->on(stream);
+
+  auto device_input_table = device_table::create(key_columns, stream);
+  auto comp = row_equality_comparator<true>(*device_input_table, true);
+
+  gdf_index_type* result_end;
+  if (keep == duplicate_keep_option::KEEP_FIRST ) {
+      result_end = thrust::unique_copy(exec, 
+              sorted_indices.begin(),
+              sorted_indices.end(),
+              unique_indices.data().get(),
+              comp
+              );
+  } else if (keep == duplicate_keep_option::KEEP_LAST) {
+      result_end = thrust::unique_copy(exec, 
+              sorted_indices.rbegin(),
+              sorted_indices.rend(),
+              unique_indices.data().get(),
+              comp
+              );
+  } else {
+      result_end = non_duplicates_copy(exec, 
+              sorted_indices.begin(),
+              sorted_indices.end(),
+              unique_indices.data().get(),
+              comp
+              );
+
+  }
+  // reorder unique indices
+  thrust::sort(exec, unique_indices.data().get(), result_end);
```

Comment:
Is this sort necessary?

Reply:
Yes. This is to get the original order back. So, the output table will be in original order as input table.

Algorithm:
- Sort the indices using row comparator
- drop duplicate indices (using stencil comparator)
- sort the indices (need to get original order - or else output table will be sorted) 
- gather these indices from input table to output table.

Reply:
But there's no need to get the original order back.

```
cudf::table drop_duplicates(const cudf::table& input_table,
                            const cudf::table& key_columns,
                            const duplicate_keep_option keep);

```

The reason for separating the `input_table` from the `key_columns` is it allows the user to pass in an Index column and re-materialize the original order if that's what they want. 

There's no reason to _always_ re-materialize the original order because that's an expensive operation.

Reply:
pandas drop_duplicates returns the original order without duplicates. so I maintained the same behavior here. I removed the resizing. but the last sort is not expensive operation. It sorts only the indices (not based on rows). (To clarify further, during the entire algorithm, the columns are never sorted. Only the indices are sorted.)

Reply:
Sorting a billion integers is _very_ expensive in comparison to doing nothing at all.

I understand what Pandas does. You're enforcing Pandas behavior at the C++ layer _always_. You're not leaving any option for a libcudf user to opt out of maintaining the original order. Python is not the only consumer of libcudf, and as such, we should _enable_ Pandas behavior, but not _force_ it.  

_IF_ the user want's the original order maintained, then they can pass in an additional column of Indices and do a `gather` on that column.

For example, if I want to drop the duplicates from a single column and maintain the original order:

```
index: [0, 1, 2, 3, 4]
input:  [5, 7, 5, 7, 4]
```

My input table would be `[index, input]`, and my `key_columns` would be `[input]`. Let's say we do `KEEP_FIRST`.

A possible output (without maintaining original order):
```
index: [4, 1, 0]
input:  [4, 7, 5]
```

Now, the _user_ is responsible for re-ordering the `input` according to the `index`, i.e.:

```
index: [0, 1, 4]
input:  [5, 7, 4]
```

Reply:
Either this, or get rid of the input/key column all together and just add an input flag to control whether or not the original order should be preserved.

Reply:
Removed the sort and updated the tests.